lib/strings.nix

```diff
@@ -545,12 +545,19 @@ rec {
        nameFromURL "https://nixos.org/releases/nix/nix-1.7/nix-1.7-x86_64-linux.tar.bz2" "_"
        => "nix-1.7-x86"
   */
-  nameFromURL = url: sep:
-    let
-      components = splitString "/" url;
-      filename = lib.last components;
-      name = head (splitString sep filename);
-    in assert name != filename; name;
+  nameFromURL = url: sep: head (splitString sep (baseNameOf url));
+
+  /* Extract name from URL, with query component or tailing slash striped.
+
+     Example:
+       nameFromURLStripQuery "https://bug-220081-attachments.webkit.org/attachment.cgi?id=416707&action=diff&format=raw"
+       => "attachment.cgi"
+       nameFromURLStripQuery "http://git.savannah.gnu.org/cgit/mailutils.git/patch/?id=37713b42a501892469234b90454731d8d8b7a3e6"
+       => "patch"
+       nameFromURLStripQuery "http://git.savannah.gnu.org/cgit/mailutils.git/patch/"
+       => "patch"
+  # */
+  nameFromURLStripQuery = url: baseNameOf (removeSuffix "/" (head (splitString "?" url)));
```

Comment:
Actually why not just make `nameFromURL` do this. I don't think there's a reason anybody would want to include parameters in the derivation name